Add option_value_fields metadata

colemanw opened the pull request:
Overview
----------------------------------------
The `OptionValue` table has a number of fields which are not relevant to every `OptionGroup`, such as "description", "icon" and "color". This adds `option_value_fields` metadata about each `OptionGroup`, which was previously hard-coded in various forms.

Before
----------------------------------------
No metadata to tell what `OptionValue` fields are relevant to which `OptionGroup`.

After
----------------------------------------
Metadata exists, and is returned by APIv4 GetFields, so it's now possible to tell which field options support a "color" or an "icon", etc.

colemanw committed:
SearchKit & Afform - Add option_value_fields info to managed option groups

colemanw commented on CRM/Admin/Form/Options.php:
```diff
@@ -174,11 +178,12 @@ public function buildQuickForm() {
     }
-
-    if (in_array($this->_gName, ['activity_status', 'case_status'])) {
+    if ($optionGroup['shape'] && in_array('color', $optionGroup['shape'])) {
```
This is the upshot of the change for our forms, we no longer need to hard-code a list of `OptionGroup`s that support "color" or "icon" we can rely on the metadata.

eileenmcnaughton commented:
I tested this & the field is correctly added / populated on upgrade and can be accessed via the api.

When adding an activity type I was offered the icon (but not colour) option before and after

My one reservation is the field name 'shape' wasn't meaningful to me - and we will be living with it for a while - @seamuslee001 @totten how do you feel about the field name 'shape'?

seamuslee001 commented:
I wonder if properties might be a better name or fields maybe?

colemanw commented:
I got the idea from PHPStorm which uses [Array Shapes](https://blog.jetbrains.com/phpstorm/2022/02/phpstorm-2022-1-eap-3/), (which are very cool by the way!) where the "shape" describes what keys an array is expected to have. It's a slightly odd term but I still can't think of a better one.

eileenmcnaughton commented:
@colemanw array shapes do look pretty cool - they differ a bit from this in that this is just a flat list of fields supported for the option group whereas array shapes describes an array whose 'shape' could vary siginifcantly (depth, string types)

I like them though

colemanw commented:
I don't want this PR to stall on a bike-sheddy thing. I'm willing to change the name to something clearer. How about `option_value_fields`? That's about as literal as you can get to describe this column's purpose.

eileenmcnaughton commented:
I'm OK with that -

seamuslee001 commented:
Sounds fine to me

colemanw committed:
Add option_value_fields metadata

This allows an optionGroup to declare which columns of the OptionValue table it uses.
Previously this was hard-coded into the form.

colemanw commented:
Merging per tag & agreement about the name.

eileenmcnaughton commented:
@colemanw thanks - I was going to give it one more r-run - but didn't see any issues with the code
